### PR TITLE
Avoiding resampling for identity transforms with integer translations

### PR DESCRIPTION
**Pull request by gkiar**
- Addresses #1551 
- Handles special case in image resampling when integer translations are accompanied by an identity affine transform in which the resampling can be fully performed by copping and padding of the source image into the target container.
- Moved `pad_img` into `nilearn.image` as it can be a useful utility outside of just testing
- Added test for `crop_img` and `pad_img`.

cc: @GaelVaroquaux

### CODE REVIEW

**Comment by GaelVaroquaux** [comment]
You have a bunch of test failures on travis.

**Comment by gkiar** [comment]
Yep - on it :) I wasn't able to run all the tests locally because surface things were being weird and seg-faulting on what I believe is untouched code... (??), but running and resolving the failures now. Will ping you when I believe it's ready

**Comment by gkiar** [comment]
@GaelVaroquaux Ready for your review I believe - while Appveyor and Circle are pending, the tests passed on Travis.

**Comment by kchawla-pi** [comment]
The error on Appveyor is non-trivial.

**Comment by GaelVaroquaux** [comment]
We're getting failures on appveyor that are related to this PR:
https://ci.appveyor.com/project/nilearn-ci/nilearn/builds/23942220

Maybe it's the tolerance of the test that is set too low.

File: doc/whats_new.rst [hunk 1]
```diff
@@ -4,6 +4,11 @@
 NEW
 ---
 
+- Optimization to image resampling
+  :func:`nilearn.image.resample_img` has been optimized to pad rather than
+  resample images in the special case when there is only a translation
+  between two spaces. This is a common case in :class:`nilearn.input_data.NiftiMasker`
+  when using the `mask_strategy="template` option for brains in MNI space.
```
Comment (GaelVaroquaux) [inline doc/whats_new.rst]:
```suggestion
  when using the `mask_strategy="template"` option for brains in MNI space.
```

File: nilearn/image/image.py [hunk 2]
```diff
@@ -364,16 +376,53 @@ def crop_img(img, rtol=1e-8, copy=True):
...
+        end = np.minimum(end + 1, data.shape[:3])
+
+    slices = [slice(s, e) for s, e in zip(start, end)][:3]
+    cropped_im = _crop_img_to(img, slices, copy=copy)
+    return cropped_im if not return_offset else (cropped_im, tuple(slices))
 
-    slices = [slice(s, e) for s, e in zip(start, end)]
 
-    return _crop_img_to(img, slices, copy=copy)
+def pad_img(array, *args):
```
Comment (GaelVaroquaux) [inline nilearn/image/image.py]: I would rather avoid the *args pattern and use "pad_sizes" as an explicit required argument:
```suggestion
def pad_img(array, pad_sizes):
```

Comment (GaelVaroquaux) [reply nilearn/image/image.py]: Also, if the input is an array, the function should be called "pad_array". But in this case, it should also probably be made private: "_pad_array"

Comment (gkiar) [reply nilearn/image/image.py]: Agreed - I just inherited this from the previously written utility I copied over. :)

**Comment by gkiar** [comment]
Regarding the issue: after just poking around, I notice that the test which fails is not touching the new code in this PR, but the old/standard resampling function (since the affine transform is not `np.eye(3)`). I verified this with debug statements locally...

I don't have a windows machine to test this on, but am happy to debug if you have suggestions for how to approach it. Thanks for the thorough review!

**Comment by gkiar** [comment]
Update: using the same version of `scipy` that is used in the appveyor build, `0.19.1`, I was able to reproduce the bug locally.

File: nilearn/image/__init__.py [hunk 3]
```diff
@@ -15,5 +15,5 @@
            'smooth_img', 'crop_img', 'mean_img', 'reorder_img',
            'swap_img_hemispheres', 'concat_imgs', 'copy_img',
            'index_img', 'iter_img', 'new_img_like', 'threshold_img',
-           'math_img', 'load_img', 'clean_img',
+           'math_img', 'load_img', 'clean_img', 'pad_array',
```
Comment (GaelVaroquaux) [inline nilearn/image/__init__.py]: Do we want to make pad_array public? I am not sure. Functions in this module are to manipulate images, and not arrays.

**Comment by kchawla-pi** [comment]
Code coverage has dropped. Are we missing some test or is this trivial?

File: nilearn/image/tests/test_resampling.py [hunk 4]
```diff
@@ -116,6 +86,7 @@ def test_downsample():
     affine = np.eye(4)
     rot_img = resample_img(Nifti1Image(data, affine),
                            target_affine=2 * affine, interpolation='nearest')
+    # import pdb; pdb.set_trace()
```
Comment (GaelVaroquaux) [inline nilearn/image/tests/test_resampling.py]: This line should probably be removed.

File: nilearn/image/tests/test_resampling.py [hunk 5]
```diff
@@ -132,6 +103,7 @@ def test_downsample():
                                target_affine=2 * affine,
                                interpolation='nearest',
                                copy=copy)
+        print(downsampled - rot_img.get_data()[:x, :y, :z, ...])
```
Comment (GaelVaroquaux) [inline nilearn/image/tests/test_resampling.py]: This line should probably be removed.

File: nilearn/image/tests/test_resampling.py [hunk 6]
```diff
@@ -558,6 +530,44 @@ def test_resample_to_img():
     np.testing.assert_almost_equal(downsampled,
                                    result_img.get_data()[:x, :y, :z, ...])
 
+def test_crop():
```
Comment (GaelVaroquaux) [inline nilearn/image/tests/test_resampling.py]: Maybe a small comment explaining the logic of this test. It helps future developers.

File: nilearn/image/tests/test_resampling.py [hunk 7]
```diff
@@ -558,6 +530,44 @@ def test_resample_to_img():
...
+    source_affine = np.eye(4)
+    source_affine[:, 3] = np.append(np.random.randint(0, 4, 3), 1)
+    source_data = rand_gen.random_sample(source_shape)
+    source_img = Nifti1Image(source_data, source_affine)
+
+    target_shape = (11, 10, 9)
+    target_data = np.zeros(target_shape)
+    target_affine = source_affine
+    target_affine[:3, 3] -= 3  # make the affine 3 larger in x, y, z
+    target_data[3:9, 3:7, 3:9] = source_data  # put the data at the offset loca
```
Comment (GaelVaroquaux) [inline nilearn/image/tests/test_resampling.py]:
```suggestion
    target_data[3:9, 3:7, 3:9] = source_data  # put the data at the offset location
```

File: nilearn/image/tests/test_resampling.py [hunk 8]
```diff
@@ -558,6 +530,44 @@ def test_resample_to_img():
...
+    source_shape = (6, 4, 6)
+    source_affine = np.eye(4)
+    source_affine[:, 3] = np.append(np.random.randint(0, 4, 3), 1)
+    source_data = rand_gen.random_sample(source_shape)
+    source_img = Nifti1Image(source_data, source_affine)
+
+    target_shape = (11, 10, 9)
+    target_data = np.zeros(target_shape)
+    target_affine = source_affine
+    target_affine[:3, 3] -= 3  # make the affine 3 larger in x, y, z
```
Comment (GaelVaroquaux) [inline nilearn/image/tests/test_resampling.py]:
```suggestion
    target_affine[:3, 3] -= 3  # add an offset of 3 in x, y, z
```

**Comment by GaelVaroquaux** [comment]
I believe that the decrease in test coverage is a glitch of codecov (it tends to do this).

File: nilearn/image/tests/test_resampling.py (hunk 4 above)
Comment (gkiar) [reply nilearn/image/tests/test_resampling.py]: yep - thanks :)

File: nilearn/image/__init__.py (hunk 3 above)
Comment (gkiar) [reply nilearn/image/__init__.py]: I'll make private, I think you're right

**Comment by gkiar** [comment]
Great - addressed comments :)

**Comment by GaelVaroquaux** [comment]
pyflakes has a warning that worries me:

https://travis-ci.org/nilearn/nilearn/jobs/522125937#L464

```
nilearn/image/image.py:406:30: F821 undefined name 'args'
                         len(args))
```

**Comment by gkiar** [comment]
Just noticed as well - when renaming a parameter from `args` to `pad_sizes` using the online comment we didn't catch all instances in the function. Just resolved.

**Comment by GaelVaroquaux** [comment]
CircleCI is taking for ever. Merging this in the interest of moving forward.

Hurray!! :rocket:

**Pull request merged**